spec/params/params.spec.ts

```diff
@@ -0,0 +1,212 @@
+import * as params from "../../src/params";
+
+describe("Params spec extraction", () => {
+  it("converts Expressions in the param default to strings", () => {
+    const bar = params.defineInt("BAR");
+    expect(
+      params.defineString("FOO", { default: bar.notEquals(22).then("asdf", "jkl;") }).toSpec()
+        .default
+    ).to.equal(`{{ params.BAR != 22 ? "asdf" : "jkl;" }}`);
+  });
+});
+
+describe("Params value extraction", () => {
+  beforeEach(() => {
+    process.env.A_STRING = "asdf";
+    process.env.SAME_STRING = "asdf";
+    process.env.DIFF_STRING = "jkl;";
+    process.env.AN_INT = "-11";
+    process.env.SAME_INT = "-11";
+    process.env.DIFF_INT = "22";
+    process.env.PI = "3.14159";
+    process.env.TRUE = "true";
+    process.env.FALSE = "false";
+  });
+
+  afterEach(() => {
+    params.clearParams();
+    delete process.env.A_STRING;
+    delete process.env.SAME_STRING;
+    delete process.env.DIFF_STRING;
+    delete process.env.AN_INT;
+    delete process.env.SAME_INT;
+    delete process.env.DIFF_INT;
+    delete process.env.TRUE;
+    delete process.env.PI;
+    delete process.env.TRUE;
+    delete process.env.FALSE;
+  });
+
+  it("extracts identity params from the environment", () => {
+    const strParam = params.defineString("A_STRING");
+    expect(strParam.value()).to.equal("asdf");
+
+    const intParam = params.defineInt("AN_INT");
+    expect(intParam.value()).to.equal(-11);
+
+    const boolParam = params.defineBoolean("TRUE");
+    expect(boolParam.value()).to.be.true;
+
+    const floatParam = params.defineFloat("PI");
+    expect(floatParam.value()).to.equal(3.14159);
+
+    const falseParam = params.defineBoolean("FALSE");
+    expect(falseParam.value()).to.be.false;
+  });
+
+  it("falls back on the javascript zero values in case of type mismatch", () => {
+    const stringToInt = params.defineInt("A_STRING");
+    expect(stringToInt.value()).to.equal(0);
+
+    const stringToBool = params.defineBoolean("A_STRING");
+    expect(stringToBool.value()).to.equal(false);
+  });
+
+  it("returns a boolean value for Comparison expressions", () => {
+    const str = params.defineString("A_STRING");
+    const sameStr = params.defineString("SAME_STRING");
+    const diffStr = params.defineString("DIFF_STRING");
+    expect(str.equals(sameStr).value()).to.be.true;
+    expect(str.equals("asdf").value()).to.be.true;
+    expect(str.equals(diffStr).value()).to.be.false;
+    expect(str.equals("jkl;").value()).to.be.false;
+
+    const int = params.defineInt("AN_INT");
+    const sameInt = params.defineInt("SAME_INT");
+    const diffInt = params.defineInt("DIFF_INT");
+    expect(int.equals(sameInt).value()).to.be.true;
+    expect(int.equals(-11).value()).to.be.true;
+    expect(int.equals(diffInt).value()).to.be.false;
+    expect(int.equals(22).value()).to.be.false;
+  });
+
+  it("can use all the comparison operators when explicitly requested", () => {
+    const jkl = params.defineString("DIFF_STRING");
```

Comment:
does comparison operator support comparing params?

e.g. does this work?

```
const asdf = params.defineString("A_STRING")
expect(jkl.cmp(">", asdf).value()).to.be.true;
```

Reply:
It does (that was the impact of changing the `rhs` type in a ComparisonExpression to be `rhs: T | Expression<T>`) but you're right in that I should add tests